Feature/enable skbuild (#668)
glotzerlab/freud: CMakeLists.txt

> cmake_minimum_required(VERSION 3.10)

Reviewer: I've seen recommendations to use at least 3.14 for "modern" features. I used 3.16 in #661.

Author: Right, on the flip side I don't know how onerous that is. 3.14 did only come out last year, I assume that for instance Ubuntu or Arch package managers for anything other than the newest OS versions might not have a sufficiently new CMake. Just want to make sure we're not causing problems for installation on e.g. any of our clusters like Comet.

Author: To be clear: you are absolutely right that I am using features that require 3.14 (all the `target_includes` that I'm doing don't exist in 3.10 AFAIK). I just want to make sure that won't be a problem for our HPC user base. HOOMD is still requiring only 3.9, although I don't know if that's just a legacy thing or planned to change for the v3 release.

Reviewer 2: HOOMD v2.x builds with CMake 2.8.10.1 or newer. HOOMD v3.x will bump that requirement to 3.9. 3.9 was the first version to have a functioning findMPI that uses modern practices. I suggest supporting the oldest version you can without needing to implement excessive workarounds. This makes the software accessible to more users, many of whom do not update their software stacks that regularly.

Reviewer: CMake version pi.0 (3.14.0) was [announced on Pi Day 2019](https://blog.kitware.com/cmake-3-14-0-available-for-download/). I'm fine with matching HOOMD's requirement of CMake 3.9 as long as it doesn't require significant workarounds. We provide conda-forge builds and/or wheels for freud on multiple platforms and architectures, so far fewer users will build from source compared to HOOMD. Also since CMake is installable with pip or conda, I think it would be acceptable to have a version requirement 3.14 that is now >18 months old.

Author: With minor modifications we're able to support CMake >= 3.3. I modified CI to test against 3.6.3, which is the oldest version available on PyPI.

cmake_minimum_required(VERSION 3.10)

project(freud)

set(DEFAULT_BUILD_TYPE "Release")

# Set a default build type if none was specified
if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    message(STATUS "Setting build type to '${DEFAULT_BUILD_TYPE}' since none was specified.")
    set(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE
        STRING "Choose the type of build." FORCE)
    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
endif()

> set(CMAKE_CXX_STANDARD 11)

Reviewer 2: Can you bump this to 14? TBB will be requiring C++14 soon.

Author: I don't see any reason that we can't. @bdice I assume you would be happy with these changes? That would also allow us to use at least some of the newer algorithms in the standard library :)

Reviewer: ```suggestion
set(CMAKE_CXX_STANDARD 14)
```

set(CMAKE_CXX_STANDARD_REQUIRED ON)

> set(CMAKE_POSITION_INDEPENDENT_CODE ON)

Reviewer 2: This shouldn't be necessary. CMake should use the right compile flags depending on the type of library being built.

Author: Got it, I had issues with it early on but it might have been because I was misusing OBJECT and SHARED in my early iterations. It does indeed seem to work correctly on shared libraries without this.

# CMake available on PyPI (which is 3.6.3).
cmake_minimum_required(VERSION 3.3.0)

project(freud)

set(DEFAULT_BUILD_TYPE "Release")

# Set a default build type if none was specified
if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    message(STATUS "Setting build type to '${DEFAULT_BUILD_TYPE}' since none was specified.")
    set(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE
        STRING "Choose the type of build." FORCE)
    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
endif()

set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
# This setting appears to be required by older versions of CMake for object libraries.
set(CMAKE_POSITION_INDEPENDENT_CODE ON)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/CMake")
find_package(TBB REQUIRED)

# Fail fast if users have not cloned submodules.
if(NOT WIN32)
  string(ASCII 27 Esc)
  set(Red         "${Esc}[31m")
  set(DefaultColor "${Esc}[m")
endif()
set(submodules
    Eigen
    fsph
    voro++)
foreach (submodule ${submodules})
    if( NOT EXISTS "${PROJECT_SOURCE_DIR}/extern/${submodule}/.git" )

>         message( FATAL_ERROR "${Red}The git submodules are not available. Please run git submodule update --init --recursive.${DefaultColor}"

Reviewer: ```suggestion
        message( FATAL_ERROR "${Red}The git submodule ${submodule} is not available. Please run git submodule update --init --recursive.${DefaultColor}"
```

Author: The loop will terminate as soon as any one submodule fails, so specifying names will be misleading.

Reviewer: But isn't it useful to know _which_ submodule raised the error? 🤨 The same solution applies regardless, but it seems more helpful than not identifying the submodule name.

Author: I disagree, because almost nobody who's going to run into this issue in practice runs `git submodule update` on specific submodules. The most common error is not initializing submodules at all, and the people making that mistake are also the most likely to see this. Under those circumstances I think it's more confusing to see a single name, especially since the recommended solution doesn't specify a submodule. If you were to error on a specific submodule, the recommended solution should also be to update only that submodule.

Reviewer: We disagree but I don't feel strongly, so you win. 😉

# oldest version of CMake that this build script has been tested with. However,
# the oldest CMake actively tested against will be determined by the oldest
# CMake available on PyPI (which is 3.6.3).
cmake_minimum_required(VERSION 3.3.0)

project(freud)

set(DEFAULT_BUILD_TYPE "Release")

# Set a default build type if none was specified
if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    message(STATUS "Setting build type to '${DEFAULT_BUILD_TYPE}' since none was specified.")
    set(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE
        STRING "Choose the type of build." FORCE)
    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
endif()

set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
# This setting is required when building shared libraries from object
# libraries. CMake intentionally chooses not to infer such information, so we
# should expect to specify this.
# https://stackoverflow.com/questions/50600708/combining-cmake-object-libraries-with-shared-libraries
set(CMAKE_POSITION_INDEPENDENT_CODE ON)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/CMake")

> include(${CMAKE_CURRENT_LIST_DIR}/CMake/freud-macros.cmake)

Reviewer 3: I think it would be more modular to call `add_subdirectory(CMake)` here, and then in that directory's CMakeLists have these two lines.

Author: Good call, I'll make that change.
